# 00 — Data Cleaning & Feature Engineering

## Overview
This notebook prepares the raw ATP match data for analysis across all subsequent notebooks in this project.

## Data Source
- **Dataset:** Jeff Sackmann's ATP Match Data
- **Source:** https://github.com/JeffSackmann/tennis_atp
- **Years used:** 2022, 2023, 2024

## Why these years?
The 2020 and 2021 seasons were significantly disrupted by the COVID-19 pandemic:
- Wimbledon 2020 was cancelled entirely
- Travel restrictions prevented several players from competing
- Tournaments were played without crowds, affecting player performance
- The competitive environment was different and less representative of normal conditions

Using 2022–2024 ensures we are working with **clean, consistent, and comparable data** from fully operational seasons.

## What this notebook does
1. Loads and combines raw CSV files into a single dataframe
2. Explores the raw data to understand its structure
3. Handles missing values with documented decisions
4. Fixes data types (date formatting)
5. Engineers new features for analysis and modelling
6. Saves the cleaned dataset to `data/processed/`

## Output
- `data/processed/matches_cleaned.csv` — used by all analysis notebooks
- `data/processed/matches_cleaned.xlsx` — demonstrates Excel compatibility

In [26]:
# Importing Libraries
import pandas as pd
import numpy as np
import os

# Display settings
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 100)

In [27]:
# Loading raw data
raw_path = '../data/raw/'

df_2022 = pd.read_csv(os.path.join(raw_path, 'atp_matches_2022.csv'))
df_2023 = pd.read_csv(os.path.join(raw_path, 'atp_matches_2023.csv'))
df_2024 = pd.read_csv(os.path.join(raw_path, 'atp_matches_2024.csv'))

# Combine all three years into one dataframe
df = pd.concat([df_2022, df_2023, df_2024], ignore_index=True)

print(f"2022: {len(df_2022)} matches")
print(f"2023: {len(df_2023)} matches")
print(f"2024: {len(df_2024)} matches")
print(f"Total: {len(df)} matches combined")

2022: 2917 matches
2023: 2986 matches
2024: 3076 matches
Total: 8979 matches combined


In [28]:
df.head()

,tourney_id,tourney_name,surface,draw_size,tourney_level,tourney_date,match_num,winner_id,winner_seed,winner_entry,winner_name,winner_hand,winner_ht,winner_ioc,winner_age,loser_id,loser_seed,loser_entry,loser_name,loser_hand,loser_ht,loser_ioc,loser_age,score,best_of,round,minutes,w_ace,w_df,w_svpt,w_1stIn,w_1stWon,w_2ndWon,w_SvGms,w_bpSaved,w_bpFaced,l_ace,l_df,l_svpt,l_1stIn,l_1stWon,l_2ndWon,l_SvGms,l_bpSaved,l_bpFaced,winner_rank,winner_rank_points,loser_rank,loser_rank_points
0,2022-8888,Atp Cup,Hard,16,A,20220103,300,200000,NaN,NaN,Felix Auger Aliassime,R,193.0,CAN,21.4,105138,NaN,NaN,Roberto Bautista Agut,R,183.0,ESP,33.7,7-6(3) 6-3,3,F,129.0,15.0,6.0,78.0,51.0,38.0,14.0,11.0,10.0,11.0,0.0,2.0,70.0,50.0,32.0,7.0,10.0,3.0,5.0,11.0,3308.0,19.0,2260.0
1,2022-8888,Atp Cup,Hard,16,A,20220103,299,133430,NaN,NaN,Denis Shapovalov,L,185.0,CAN,22.7,105807,NaN,NaN,Pablo Carreno Busta,R,188.0,ESP,30.4,6-4 6-3,3,F,98.0,7.0,2.0,78.0,49.0,34.0,16.0,10.0,8.0,9.0,1.0,0.0,50.0,33.0,21.0,8.0,9.0,3.0,6.0,14.0,2475.0,20.0,2230.0
2,2022-8888,Atp Cup,Hard,16,A,20220103,298,105138,NaN,NaN,Roberto Bautista Agut,R,183.0,ESP,33.7,128034,NaN,NaN,Hubert Hurkacz,R,196.0,POL,24.8,7-6(6) 2-6 7-6(5),3,SF,164.0,1.0,2.0,96.0,64.0,50.0,20.0,16.0,1.0,4.0,24.0,3.0,120.0,80.0,62.0,20.0,16.0,6.0,7.0,19.0,2260.0,9.0,3706.0
3,2022-8888,Atp Cup,Hard,16,A,20220103,297,105807,NaN,NaN,Pablo Carreno Busta,R,188.0,ESP,30.4,126591,NaN,NaN,Jan Zielinski,R,180.0,POL,25.1,6-2 6-1,3,SF,53.0,6.0,0.0,45.0,33.0,25.0,8.0,8.0,0.0,0.0,2.0,1.0,38.0,27.0,17.0,1.0,7.0,4.0,8.0,20.0,2230.0,860.0,18.0
4,2022-8888,Atp Cup,Hard,16,A,20220103,296,106421,NaN,NaN,Daniil Medvedev,R,198.0,RUS,25.8,200000,NaN,NaN,Felix Auger Aliassime,R,193.0,CAN,21.4,6-4 6-0,3,SF,68.0,6.0,4.0,41.0,25.0,22.0,10.0,8.0,0.0,0.0,6.0,2.0,48.0,35.0,22.0,4.0,8.0,3.0,7.0,2.0,8640.0,11.0,3308.0


In [29]:
df.shape

(8979, 49)

In [30]:
print(df.columns.tolist())

['tourney_id', 'tourney_name', 'surface', 'draw_size', 'tourney_level', 'tourney_date', 'match_num', 'winner_id', 'winner_seed', 'winner_entry', 'winner_name', 'winner_hand', 'winner_ht', 'winner_ioc', 'winner_age', 'loser_id', 'loser_seed', 'loser_entry', 'loser_name', 'loser_hand', 'loser_ht', 'loser_ioc', 'loser_age', 'score', 'best_of', 'round', 'minutes', 'w_ace', 'w_df', 'w_svpt', 'w_1stIn', 'w_1stWon', 'w_2ndWon', 'w_SvGms', 'w_bpSaved', 'w_bpFaced', 'l_ace', 'l_df', 'l_svpt', 'l_1stIn', 'l_1stWon', 'l_2ndWon', 'l_SvGms', 'l_bpSaved', 'l_bpFaced', 'winner_rank', 'winner_rank_points', 'loser_rank', 'loser_rank_points']


In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8979 entries, 0 to 8978
Data columns (total 49 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   tourney_id          8979 non-null   object 
 1   tourney_name        8979 non-null   object 
 2   surface             8926 non-null   object 
 3   draw_size           8979 non-null   int64  
 4   tourney_level       8979 non-null   object 
 5   tourney_date        8979 non-null   int64  
 6   match_num           8979 non-null   int64  
 7   winner_id           8979 non-null   int64  
 8   winner_seed         3764 non-null   float64
 9   winner_entry        1378 non-null   object 
 10  winner_name         8979 non-null   object 
 11  winner_hand         8979 non-null   object 
 12  winner_ht           8934 non-null   float64
 13  winner_ioc          8979 non-null   object 
 14  winner_age          8977 non-null   float64
 15  loser_id            8979 non-null   int64  
 16  loser_

In [32]:
# Check missing values
missing = pd.DataFrame({
    'missing_count': df.isnull().sum(),
    'missing_percent': (df.isnull().sum() / len(df) * 100).round(2)
})

missing = missing[missing['missing_count'] > 0].sort_values('missing_percent', ascending=False)

print(missing)


                    missing_count  missing_percent
winner_entry                 7601            84.65
loser_entry                  6912            76.98
loser_seed                   6748            75.15
winner_seed                  5215            58.08
minutes                       750             8.35
w_SvGms                       404             4.50
l_SvGms                       404             4.50
w_2ndWon                      403             4.49
l_1stWon                      403             4.49
l_1stIn                       403             4.49
l_svpt                        403             4.49
l_df                          403             4.49
l_ace                         403             4.49
w_bpFaced                     403             4.49
w_bpSaved                     403             4.49
w_svpt                        403             4.49
w_1stWon                      403             4.49
w_1stIn                       403             4.49
w_df                          4

In [33]:
# Dropping columns that are irrelevant to this analysis
cols_not_relevant = [
    'winner_ht', 'loser_ht',
    'winner_hand', 'loser_hand',
    'winner_ioc', 'loser_ioc',
    'tourney_id',
    'match_num',
    'winner_id', 'loser_id',
    'w_ace', 'w_df', 'w_svpt',
    'w_1stIn', 'w_1stWon', 'w_2ndWon',
    'w_SvGms', 'w_bpSaved', 'w_bpFaced',
    'l_ace', 'l_df', 'l_svpt',
    'l_1stIn', 'l_1stWon', 'l_2ndWon',
    'l_SvGms', 'l_bpSaved', 'l_bpFaced'
]

df = df.drop(columns=cols_not_relevant)
print(df.shape)

(8979, 21)


In [34]:
# Dropping columns with too much missing data
cols_high_missing = [
    'winner_entry', 'loser_entry',      # 77-85% missing
    'winner_seed', 'loser_seed',        # 58-75% missing
]

df = df.drop(columns=cols_high_missing)
print(f"Columns after dropping high-missing columns: {df.shape[1]}")
print(df.columns.tolist())

Columns after dropping high-missing columns: 17
['tourney_name', 'surface', 'draw_size', 'tourney_level', 'tourney_date', 'winner_name', 'winner_age', 'loser_name', 'loser_age', 'score', 'best_of', 'round', 'minutes', 'winner_rank', 'winner_rank_points', 'loser_rank', 'loser_rank_points']


In [35]:
# Dropping rows with critical missing values
before = len(df)

# Drop rows where ranking data is missing — we can't analyse without it
df = df.dropna(subset=['winner_rank', 'loser_rank'])

# Drop rows where surface is missing — needed for surface analysis
df = df.dropna(subset=['surface'])

# Drop rows where age is missing — only 2-3 rows, not worth keeping
df = df.dropna(subset=['winner_age', 'loser_age'])

after = len(df)

print(f"Rows before: {before}")
print(f"Rows after: {after}")
print(f"Rows dropped: {before - after}")

Rows before: 8979
Rows after: 8827
Rows dropped: 152


In [36]:
# Checking 'minutes' column
missing_minutes = df[df['minutes'].isnull()]
print(f"Total missing minutes: {len(missing_minutes)}")
print(f"\nSample of scores where minutes is missing:")
print(missing_minutes['score'].value_counts().head(20))

Total missing minutes: 660

Sample of scores where minutes is missing:
score
6-3 6-4        19
W/O            18
6-2 6-2        15
6-4 6-4        14
6-4 6-3        13
6-4 6-2        12
6-1 6-4        11
6-2 6-1         9
6-4 7-5         8
6-2 6-4         7
6-3 6-2         7
7-6(3) 6-2      6
7-5 6-3         6
6-3 6-3         6
6-3 7-5         5
6-1 6-3         5
6-3 6-1         5
6-1 6-1         5
6-3 6-4 6-3     5
6-1 6-2         5
Name: count, dtype: int64


In [37]:
# Check for retirements and walkovers in missing minutes
# Check for walkovers
walkovers = missing_minutes[missing_minutes['score'].str.contains('W/O', na=False)]
print(f"Walkovers: {len(walkovers)}")

# Check for retirements
retirements = missing_minutes[missing_minutes['score'].str.contains('RET', na=False)]
print(f"Retirements: {len(retirements)}")

# Everything else — completed matches with missing duration
completed = missing_minutes[
    ~missing_minutes['score'].str.contains('W/O|RET', na=False)
]
print(f"Completed matches with missing duration: {len(completed)}")

Walkovers: 18
Retirements: 9
Completed matches with missing duration: 633


In [38]:
# Explore median duration by surface and best_of
median_by_group = df[df['minutes'] > 0].groupby(['surface', 'best_of'])['minutes'].median()
print("Median match duration by surface and best_of:")
print(median_by_group)

Median match duration by surface and best_of:
surface  best_of
Clay     3          107.0
         5          169.5
Grass    3          100.0
         5          151.0
Hard     3          101.0
         5          159.0
Name: minutes, dtype: float64


In [39]:
# Imputing missing values in 'minutes' column
# Fill walkovers with 0 — no time spent on court
df.loc[df['score'].str.contains('W/O', na=False) & df['minutes'].isnull(), 'minutes'] = 0

# Fill retirements with 60% of median by surface and best_of
retirement_mask = df['score'].str.contains('RET', na=False) & df['minutes'].isnull()
df.loc[retirement_mask, 'minutes'] = df.loc[retirement_mask].apply(
    lambda row: round(median_by_group.get((row['surface'], row['best_of']), median_by_group.mean()) * 0.6),
    axis=1
)

# Fill remaining completed matches with median by surface and best_of
def fill_median(row):
    if pd.isnull(row['minutes']):
        return median_by_group.get((row['surface'], row['best_of']), median_by_group.mean())
    return row['minutes']

df['minutes'] = df.apply(fill_median, axis=1)

print(f"Missing minutes remaining: {df['minutes'].isnull().sum()}")
print(f"\nMinutes distribution after imputation:")
print(df['minutes'].describe())

Missing minutes remaining: 0

Minutes distribution after imputation:
count    8827.000000
mean      117.420641
std        42.628940
min         0.000000
25%        87.000000
50%       108.000000
75%       144.000000
max       345.000000
Name: minutes, dtype: float64


## Missing Data Imputation Strategy — Match Duration (`minutes`)

Out of **660 missing values** in the `minutes` column, we identified three distinct categories
by examining the `score` column:

| Category | Count | Imputation Strategy | Reasoning |
|---|---|---|---|
| Walkovers (`W/O`) | 18 | Fill with **0** | Player never took the court — zero time spent playing |
| Retirements (`RET`) | 9 | Fill with **60% of group median** | Match started but did not finish — assumed roughly 60% of a full match was played |
| Completed matches | 633 | Fill with **group median** | Match finished but duration was not recorded |

### Why group median instead of global median?

Rather than using a single global median, we calculated the median match duration
grouped by **surface** and **best_of** (number of sets). This accounts for two
well-known sources of variation in match duration:

- **Surface:** Clay court matches are longer due to slower rallies. Grass is the fastest surface.
- **Best of 5 vs Best of 3:** Grand Slam matches (best of 5) are on average 50–60 minutes
longer than regular tour matches (best of 3).

Using a global median would have misrepresented player workload, directly impacting
the accuracy of our fatigue monitoring analysis.

In [43]:
# Fixing Date format
# Check what the date looks like currently
print(f"Current format: {df['tourney_date'].dtype}")
print(f"Sample values: {df['tourney_date'].head().tolist()}")

# Convert from integer (20220107) to proper datetime
df['tourney_date'] = pd.to_datetime(df['tourney_date'], format='%Y%m%d')

# Verify conversion
print(f"\nNew format: {df['tourney_date'].dtype}")
print(f"Sample values: {df['tourney_date'].head().tolist()}")
print(f"\nDate range: {df['tourney_date'].min()} to {df['tourney_date'].max()}")

Current format: datetime64[ns]
Sample values: [Timestamp('2022-01-03 00:00:00'), Timestamp('2022-01-03 00:00:00'), Timestamp('2022-01-03 00:00:00'), Timestamp('2022-01-03 00:00:00'), Timestamp('2022-01-03 00:00:00')]

New format: datetime64[ns]
Sample values: [Timestamp('2022-01-03 00:00:00'), Timestamp('2022-01-03 00:00:00'), Timestamp('2022-01-03 00:00:00'), Timestamp('2022-01-03 00:00:00'), Timestamp('2022-01-03 00:00:00')]

Date range: 2022-01-03 00:00:00 to 2024-12-18 00:00:00


In [45]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8827 entries, 0 to 8977
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   tourney_name        8827 non-null   object        
 1   surface             8827 non-null   object        
 2   draw_size           8827 non-null   int64         
 3   tourney_level       8827 non-null   object        
 4   tourney_date        8827 non-null   datetime64[ns]
 5   winner_name         8827 non-null   object        
 6   winner_age          8827 non-null   float64       
 7   loser_name          8827 non-null   object        
 8   loser_age           8827 non-null   float64       
 9   score               8827 non-null   object        
 10  best_of             8827 non-null   int64         
 11  round               8827 non-null   object        
 12  minutes             8827 non-null   float64       
 13  winner_rank         8827 non-null   float64       
 1

In [47]:
df.describe().transpose()

,count,mean,min,25%,50%,75%,max,std
draw_size,8827.0,62.671576,2.0,32.0,32.0,128.0,128.0,44.5575
tourney_date,8827,2023-06-08 14:33:35.543219712,2022-01-03 00:00:00,2022-08-29 00:00:00,2023-06-19 00:00:00,2024-03-18 00:00:00,2024-12-18 00:00:00,NaN
winner_age,8827.0,26.546868,17.0,23.5,26.0,28.7,41.7,4.372423
loser_age,8827.0,26.940263,15.6,23.8,26.5,29.6,41.7,4.423114
best_of,8827.0,3.354367,3.0,3.0,3.0,3.0,5.0,0.763691
minutes,8827.0,117.420641,0.0,87.0,108.0,144.0,345.0,42.62894
winner_rank,8827.0,69.981761,1.0,17.0,44.0,84.0,1783.0,107.081257
winner_rank_points,8827.0,1830.64518,1.0,697.0,1033.0,2235.0,11830.0,1926.162866
loser_rank,8827.0,104.910842,1.0,35.0,65.0,110.0,1881.0,152.558854
loser_rank_points,8827.0,1168.122012,1.0,550.0,818.0,1258.5,11245.0,1237.18304


## Feature Engineering & Data Reshaping

This section transforms the raw match data into a player-centric format suitable
for analysis and machine learning. It involves two major steps:

### Step 1 — Reshape from match format to player format
The raw data has one row per match (winner vs loser). We reshape it to have
**two rows per match** — one for each player — giving each player their own
perspective of the match. This is essential for building an unbiased ML model
with a balanced 50/50 win/loss target variable.

### Step 2 — Feature engineering
We create the following new columns:

| Feature | Description | Used in |
|---|---|---|
| `win` | 1 if player won, 0 if lost | Notebooks 01, 04 |
| `rank_diff` | Player rank minus opponent rank | Notebooks 01, 04 |
| `opponent_rank` | Ranking of the opponent | Notebooks 01, 04 |
| `opponent_rank_points` | Rank points of the opponent | Notebook 04 |
| `match_load_7` | Matches played in last 7 days | Notebook 05 |
| `match_load_14` | Matches played in last 14 days | Notebook 05 |
| `fatigue_score_7` | Normalised 7-day match load (0-1) | Notebook 05 |
| `fatigue_score_14` | Normalised 14-day match load (0-1) | Notebook 05 |

### Why 7 and 14 days for match load?
- **7 days** captures acute fatigue — tiredness from the current tournament week
- **14 days** captures cumulative fatigue — tiredness built up over two weeks,
  relevant for Grand Slams and back-to-back tournaments

### Why a balanced 50/50 dataset is ideal
With exactly 50% wins and 50% losses, the model is forced to learn genuine
patterns rather than exploiting class imbalance. A random baseline would achieve
50% accuracy — our model should comfortably exceed this.

In [48]:
# Reshape from match format to player format
# Shared match columns (same for both players)
shared_cols = [
    'tourney_name', 'tourney_date', 'tourney_level',
    'surface', 'draw_size', 'round', 'best_of', 'minutes', 'score'
]

# Winner perspective
winner_df = df[shared_cols].copy()
winner_df['player_name'] = df['winner_name']
winner_df['player_rank'] = df['winner_rank']
winner_df['player_rank_points'] = df['winner_rank_points']
winner_df['player_age'] = df['winner_age']
winner_df['opponent_rank'] = df['loser_rank']
winner_df['opponent_rank_points'] = df['loser_rank_points']
winner_df['win'] = 1

# Loser perspective
loser_df = df[shared_cols].copy()
loser_df['player_name'] = df['loser_name']
loser_df['player_rank'] = df['loser_rank']
loser_df['player_rank_points'] = df['loser_rank_points']
loser_df['player_age'] = df['loser_age']
loser_df['opponent_rank'] = df['winner_rank']
loser_df['opponent_rank_points'] = df['winner_rank_points']
loser_df['win'] = 0

# Stack both perspectives
df_long = pd.concat([winner_df, loser_df], ignore_index=True)

# Sort by player and date — essential for match_load calculation
df_long = df_long.sort_values(['player_name', 'tourney_date']).reset_index(drop=True)

print(f"Original shape: {df.shape}")
print(f"Reshaped shape: {df_long.shape}")
print(f"\nWin distribution:")
print(df_long['win'].value_counts())
print(f"\nSample rows:")
print(df_long[['player_name', 'tourney_date', 'surface', 'player_rank', 'opponent_rank', 'win']].head(10))

Original shape: (8827, 17)
Reshaped shape: (17654, 16)

Win distribution:
win
0    8827
1    8827
Name: count, dtype: int64

Sample rows:
          player_name tourney_date surface  player_rank  opponent_rank  win
0  Abedallah Shelbayh   2023-02-20    Hard        276.0           68.0    0
1  Abedallah Shelbayh   2023-04-17    Clay        267.0          139.0    1
2  Abedallah Shelbayh   2023-04-17    Clay        267.0           33.0    0
3  Abedallah Shelbayh   2023-04-24    Clay        243.0           67.0    0
4  Abedallah Shelbayh   2023-06-26   Grass        240.0           94.0    0
5  Abedallah Shelbayh   2023-07-17   Grass        240.0          116.0    0
6  Abedallah Shelbayh   2023-11-06    Hard        215.0           83.0    1
7  Abedallah Shelbayh   2023-11-06    Hard        215.0           49.0    0
8  Abedallah Shelbayh   2023-11-27    Hard        185.0           97.0    1
9  Abedallah Shelbayh   2023-11-27    Hard        185.0           70.0    0


In [49]:
df_long.columns.tolist()

['tourney_name',
 'tourney_date',
 'tourney_level',
 'surface',
 'draw_size',
 'round',
 'best_of',
 'minutes',
 'score',
 'player_name',
 'player_rank',
 'player_rank_points',
 'player_age',
 'opponent_rank',
 'opponent_rank_points',
 'win']

In [50]:
# Creating rank_diff feature
# rank_diff = player_rank - opponent_rank
# Negative value = player is higher ranked (better) than opponent
# Positive value = player is lower ranked (worse) than opponent
df_long['rank_diff'] = df_long['player_rank'] - df_long['opponent_rank']

print(f"rank_diff sample statistics:")
print(df_long['rank_diff'].describe())
print(f"\nSample:")
print(df_long[['player_name', 'player_rank', 'opponent_rank', 'rank_diff', 'win']].head(10))

rank_diff sample statistics:
count    17654.000000
mean         0.000000
std        154.317789
min      -1791.000000
25%        -42.000000
50%          0.000000
75%         42.000000
max       1791.000000
Name: rank_diff, dtype: float64

Sample:
          player_name  player_rank  opponent_rank  rank_diff  win
0  Abedallah Shelbayh        276.0           68.0      208.0    0
1  Abedallah Shelbayh        267.0          139.0      128.0    1
2  Abedallah Shelbayh        267.0           33.0      234.0    0
3  Abedallah Shelbayh        243.0           67.0      176.0    0
4  Abedallah Shelbayh        240.0           94.0      146.0    0
5  Abedallah Shelbayh        240.0          116.0      124.0    0
6  Abedallah Shelbayh        215.0           83.0      132.0    1
7  Abedallah Shelbayh        215.0           49.0      166.0    0
8  Abedallah Shelbayh        185.0           97.0       88.0    1
9  Abedallah Shelbayh        185.0           70.0      115.0    0


In [52]:
# Creating match_load_7 and match_load_14
# For each player, count how many matches they played
# in the 7 and 14 days prior to each match (not including the match itself)

def calculate_match_load(group, days):
    load = []
    dates = group['tourney_date'].values
    for i, date in enumerate(dates):
        cutoff = date - pd.Timedelta(days=days)
        count = sum(1 for d in dates[:i] if d >= cutoff)
        load.append(count)
    return load

df_long['match_load_7'] = (
    df_long.groupby('player_name', group_keys=False)
    .apply(lambda g: pd.Series(calculate_match_load(g, 7), index=g.index), include_groups=False)
)

df_long['match_load_14'] = (
    df_long.groupby('player_name', group_keys=False)
    .apply(lambda g: pd.Series(calculate_match_load(g, 14), index=g.index), include_groups=False)
)

print(f"Match load statistics (7 days):")
print(df_long['match_load_7'].describe())
print(f"\nMatch load statistics (14 days):")
print(df_long['match_load_14'].describe())


Match load statistics (7 days):
count    17654.000000
mean         1.575677
std          1.671593
min          0.000000
25%          0.000000
50%          1.000000
75%          2.000000
max         10.000000
Name: match_load_7, dtype: float64

Match load statistics (14 days):
count    17654.000000
mean         2.440297
std          2.190944
min          0.000000
25%          1.000000
50%          2.000000
75%          4.000000
max         15.000000
Name: match_load_14, dtype: float64


In [53]:
# Creating fatigue_score columns
# Normalise match load to 0-1 scale using min-max normalisation
# 0 = fully fresh, 1 = highest recorded match load in the dataset

df_long['fatigue_score_7'] = (
    df_long['match_load_7'] / df_long['match_load_7'].max()
)

df_long['fatigue_score_14'] = (
    df_long['match_load_14'] / df_long['match_load_14'].max()
)

print(f"Fatigue score (7 day) range: {df_long['fatigue_score_7'].min():.2f} to {df_long['fatigue_score_7'].max():.2f}")
print(f"Fatigue score (14 day) range: {df_long['fatigue_score_14'].min():.2f} to {df_long['fatigue_score_14'].max():.2f}")
print(f"\nSample:")
print(df_long[['player_name', 'tourney_date', 'match_load_7', 'fatigue_score_7', 'match_load_14', 'fatigue_score_14']].head(10))

Fatigue score (7 day) range: 0.00 to 1.00
Fatigue score (14 day) range: 0.00 to 1.00

Sample:
          player_name tourney_date  match_load_7  fatigue_score_7  \
0  Abedallah Shelbayh   2023-02-20             0              0.0   
1  Abedallah Shelbayh   2023-04-17             0              0.0   
2  Abedallah Shelbayh   2023-04-17             1              0.1   
3  Abedallah Shelbayh   2023-04-24             2              0.2   
4  Abedallah Shelbayh   2023-06-26             0              0.0   
5  Abedallah Shelbayh   2023-07-17             0              0.0   
6  Abedallah Shelbayh   2023-11-06             0              0.0   
7  Abedallah Shelbayh   2023-11-06             1              0.1   
8  Abedallah Shelbayh   2023-11-27             0              0.0   
9  Abedallah Shelbayh   2023-11-27             1              0.1   

   match_load_14  fatigue_score_14  
0              0          0.000000  
1              0          0.000000  
2              1          0.066667 

In [54]:
# Final check
print(f"Final shape: {df_long.shape}")
print(f"\nColumns:")
print(df_long.columns.tolist())
print(f"\nMissing values:")
print(df_long.isnull().sum())
print(f"\nSample of final dataframe:")
print(df_long.head())

Final shape: (17654, 21)

Columns:
['tourney_name', 'tourney_date', 'tourney_level', 'surface', 'draw_size', 'round', 'best_of', 'minutes', 'score', 'player_name', 'player_rank', 'player_rank_points', 'player_age', 'opponent_rank', 'opponent_rank_points', 'win', 'rank_diff', 'match_load_7', 'match_load_14', 'fatigue_score_7', 'fatigue_score_14']

Missing values:
tourney_name            0
tourney_date            0
tourney_level           0
surface                 0
draw_size               0
round                   0
best_of                 0
minutes                 0
score                   0
player_name             0
player_rank             0
player_rank_points      0
player_age              0
opponent_rank           0
opponent_rank_points    0
win                     0
rank_diff               0
match_load_7            0
match_load_14           0
fatigue_score_7         0
fatigue_score_14        0
dtype: int64

Sample of final dataframe:
     tourney_name tourney_date tourney_level sur

In [55]:
# Save cleaned Data
# Define output path
processed_path = '../data/processed/'

# Save as CSV — used by all analysis notebooks
df_long.to_csv(os.path.join(processed_path, 'matches_cleaned.csv'), index=False)
print("Saved matches_cleaned.csv")

# Save as Excel — demonstrates Excel compatibility
df_long.to_excel(os.path.join(processed_path, 'matches_cleaned.xlsx'), index=False)
print("Saved matches_cleaned.xlsx")

print(f"\nFinal dataset: {df_long.shape[0]} rows x {df_long.shape[1]} columns")
print(f"Saved to: {processed_path}")

Saved matches_cleaned.csv
Saved matches_cleaned.xlsx

Final dataset: 17654 rows x 21 columns
Saved to: ../data/processed/


## Summary & Conclusions

This notebook successfully prepared the raw ATP match data for all subsequent analysis.

### Data decisions made

| Decision | Reasoning |
|---|---|
| Used 2022–2024 only | Avoids COVID-disrupted seasons (2020–2021) which are not representative of normal competitive conditions |
| Dropped 24 columns | Removed irrelevant, granular, or high-missing columns that add no analytical value |
| Dropped 152 rows | Removed rows with missing ranking or surface data — less than 2% of the dataset |
| Imputed minutes by surface + best_of | More accurate than global median — accounts for known variation in match duration |
| Filled walkovers with 0 | Player never took the court — zero match load |
| Filled retirements with 60% of group median | Match started but did not finish |
| Reshaped to player format | One row per player per match enables unbiased ML modelling with balanced 50/50 target |

### Output
- **17,654 rows** — two rows per match, one per player
- **21 columns** — clean, analysis-ready features
- **Zero missing values**

### Future considerations
- Nationality and travel distance between tournaments could further refine
  fatigue estimates — a player travelling from Europe to Asia faces additional
  jet lag not captured in match duration alone
- A 30-day match load window could capture season-level chronic fatigue
  in addition to the 7 and 14 day acute fatigue measures